crates/engine/tree/src/tree/memory_overlay.rs

```diff
@@ -0,0 +1,123 @@
+use super::ExecutedBlock;
+use reth_errors::ProviderResult;
+use reth_primitives::{Account, Address, BlockNumber, Bytecode, StorageKey, StorageValue, B256};
+use reth_provider::{AccountReader, BlockHashReader, StateProvider, StateRootProvider};
+use reth_trie::{updates::TrieUpdates, AccountProof};
+use revm::db::BundleState;
+
+/// A state provider that stores references to in-memory blocks along with their state as well as
+/// the historical state provider for fallback lookups.
+pub struct MemoryOverlayStateProvider<H> {
+    /// The collection of executed parent blocks.
+    in_memory: Vec<ExecutedBlock>,
+    /// Historical state provider for state lookups that are not found in in-memory blocks.
+    historical: H,
+}
+
+impl<H> MemoryOverlayStateProvider<H> {
+    /// Create new memory overlay state provider.
+    pub fn new(in_memory: Vec<ExecutedBlock>, historical: H) -> Self {
+        Self { in_memory, historical }
+    }
+}
+
+impl<H> BlockHashReader for MemoryOverlayStateProvider<H>
+where
+    H: BlockHashReader,
+{
+    fn block_hash(&self, number: BlockNumber) -> ProviderResult<Option<B256>> {
+        for block in self.in_memory.iter().rev() {
+            if block.block.number == number {
+                return Ok(Some(block.block.hash()))
+            }
+        }
+
+        self.historical.block_hash(number)
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        start: BlockNumber,
+        end: BlockNumber,
+    ) -> ProviderResult<Vec<B256>> {
+        let range = start..end;
+        let mut earliest_block_number = None;
+        let mut in_memory_hashes = Vec::new();
+        for block in self.in_memory.iter().rev() {
+            if range.contains(&block.block.number) {
+                in_memory_hashes.insert(0, block.block.hash());
+                earliest_block_number = Some(block.block.number);
+            }
+        }
+
+        let mut hashes =
+            self.historical.canonical_hashes_range(start, earliest_block_number.unwrap_or(end))?;
+        hashes.append(&mut in_memory_hashes);
+        Ok(hashes)
+    }
+}
+
+impl<H> AccountReader for MemoryOverlayStateProvider<H>
+where
+    H: AccountReader + Send,
+{
+    fn basic_account(&self, address: Address) -> ProviderResult<Option<Account>> {
+        for block in self.in_memory.iter().rev() {
+            if let Some(account) = block.execution_output.account(&address) {
+                return Ok(account)
+            }
+        }
+
+        self.historical.basic_account(address)
+    }
+}
+
+impl<H> StateRootProvider for MemoryOverlayStateProvider<H>
+where
+    H: StateRootProvider + Send,
+{
+    fn state_root(&self, bundle_state: &BundleState) -> ProviderResult<B256> {
+        todo!()
+    }
+
+    fn state_root_with_updates(
+        &self,
+        bundle_state: &BundleState,
+    ) -> ProviderResult<(B256, TrieUpdates)> {
+        todo!()
+    }
```

Comment:
do we need this https://github.com/paradigmxyz/reth/pull/8199/files?

Reply:
ye, we'll need some variant of it